src/os/bluestore/bluestore_types.h

```diff
@@ -628,56 +636,106 @@ struct bluestore_blob_t {
+        i++;
+      }
+      return i >= end;
     }
-    if (start != end) {
-      set_flag(FLAG_HAS_UNUSED);
+    if (has_unused()) {
+      uint64_t blob_unused_len = min_alloc_size;
+      ceph_assert((blob_unused_len % (sizeof(unused)*8)) == 0);
+      ceph_assert(length <= blob_unused_len);
+      uint64_t chunk_size = blob_unused_len / (sizeof(unused)*8);
+      uint64_t blob_unused_off = get_unused_off();
+      if (offset < blob_unused_off) {
+        // can happen if we are seeking a blob for re-use
+        return false;
+      }
+      uint64_t start_unused_off = offset - blob_unused_off;
+      if (start_unused_off + length > blob_unused_len) {
+        // can happen if we are seeking a blob for re-use
+        return false;
+      }
+      uint64_t start = start_unused_off / chunk_size;
+      uint64_t end = round_up_to(start_unused_off + length, chunk_size) /
+                     chunk_size;
+      auto i = start;
+      while (i < end && (unused & (1u << i))) {
+        i++;
+      }
+      return i >= end;
     }
+    return false;
   }
 
-  /// indicate that a range has (now) been used.
-  void mark_used(uint64_t offset, uint64_t length) {
-    if (has_unused()) {
+  void mark_all_unused() {
+    set_flag(FLAG_HAS_UNUSED);
+    unused = -1;
+  }
+
+  /// calculate which unused bits are affected by the input range
+  unused_t calc_bits(uint64_t offset,
+                     uint64_t length,
+                     uint32_t min_alloc_size) const {
+    unused_t u = 0;
+    if (has_legacy_unused()) {
       uint64_t blob_len = get_logical_length();
       ceph_assert((blob_len % (sizeof(unused)*8)) == 0);
       ceph_assert(offset + length <= blob_len);
       uint64_t chunk_size = blob_len / (sizeof(unused)*8);
       uint64_t start = offset / chunk_size;
       uint64_t end = round_up_to(offset + length, chunk_size) / chunk_size;
       for (auto i = start; i < end; ++i) {
-        unused &= ~(1u << i);
+        u |= 1u << i;
       }
-      if (unused == 0) {
-        clear_flag(FLAG_HAS_UNUSED);
+    }
+    if (has_unused()) {
+      uint64_t blob_unused_len = min_alloc_size;
+      ceph_assert((blob_unused_len % (sizeof(unused)*8)) == 0);
+      ceph_assert(length <= blob_unused_len);
+      uint64_t chunk_size = blob_unused_len / (sizeof(unused)*8);
+      uint64_t blob_unused_off = get_unused_off();
+      ceph_assert(offset >= blob_unused_off);
+      uint64_t start_unused_off = offset - blob_unused_off;
+      ceph_assert(start_unused_off + length <= blob_unused_len);
+      uint64_t start = start_unused_off / chunk_size;
+      uint64_t end = round_up_to(start_unused_off + length, chunk_size) /
+                     chunk_size;
+      for (auto i = start; i < end; ++i) {
+        u |= 1u << i;
       }
     }
+    return u;
+  }
+
+  /// indicate that a range has (now) been used.
+  void mark_used(uint64_t offset, uint64_t length, uint32_t min_alloc_size) {
+    auto used = calc_bits(offset, length, min_alloc_size);
```

Comment:
May be cut the path for a bit and additionally check here for (has_unused() || has_legacy_unused()) and bypass calc_bits and other stuff if possible?